Add libhvee as alternative Hyper-V driver on Windows

gbraad opened the pull request:
Fixes #3811 

Basic functionality for:
 - create
 - start
 - stop
 - delete

gbraad committed:
Use libhvee instead of the current PowerShell cmdlet implementation

This replaces the interaction of Hyper-V from our machine-drivers based
driver as used in Minikube/minishift to the libhvee library as used by
podman machine.

gbraad commented:
rebase broke the build:

```
Error: pkg\crc\preflight\preflight_checks_windows.go:14:2: cannot find package "." in:
	D:\a\crc\crc\vendor\github.com\crc-org\crc\pkg\crc\constants
Error: pkg\crc\machine\driver_windows.go:7:2: cannot find package "." in:
	D:\a\crc\crc\vendor\github.com\crc-org\crc\pkg\crc\machine\config
Error: pkg\crc\machine\driver_windows.go:8:2: cannot find package "." in:
	D:\a\crc\crc\vendor\github.com\crc-org\crc\pkg\crc\machine\libhvee
Error: pkg\crc\machine\driver_windows.go:9:2: cannot find package "." in:
	D:\a\crc\crc\vendor\github.com\crc-org\crc\pkg\drivers\libhvee
Error: pkg\crc\machine\driver_windows.go:10:2: cannot find package "." in:
	D:\a\crc\crc\vendor\github.com\crc-org\crc\pkg\libmachine
Error: pkg\crc\machine\driver_windows.go:11:2: cannot find package "." in:
	D:\a\crc\crc\vendor\github.com\crc-org\crc\pkg\libmachine\host
```

and for my local build I get:

```
go: finding module for package github.com/crc-org/crc/pkg/crc/constants
go: finding module for package github.com/crc-org/crc/pkg/crc/machine/config
go: finding module for package github.com/crc-org/crc/pkg/libmachine
go: finding module for package github.com/crc-org/crc/pkg/drivers/libhvee
go: finding module for package github.com/crc-org/crc/pkg/libmachine/host
go: finding module for package github.com/crc-org/crc/pkg/crc/machine/libhvee
pkg/crc/preflight/preflight_checks_windows.go:14:2: cannot query module due to -mod=vendor
        (Go version in go.mod is at least 1.14 and vendor directory exists.)
pkg/crc/machine/driver_windows.go:7:2: cannot query module due to -mod=vendor
        (Go version in go.mod is at least 1.14 and vendor directory exists.)
pkg/crc/machine/driver_windows.go:8:2: cannot query module due to -mod=vendor
        (Go version in go.mod is at least 1.14 and vendor directory exists.)
pkg/crc/machine/driver_windows.go:9:2: cannot query module due to -mod=vendor
        (Go version in go.mod is at least 1.14 and vendor directory exists.)
pkg/crc/machine/driver_windows.go:10:2: cannot query module due to -mod=vendor
        (Go version in go.mod is at least 1.14 and vendor directory exists.)
pkg/crc/machine/driver_windows.go:11:2: cannot query module due to -mod=vendor
        (Go version in go.mod is at least 1.14 and vendor directory exists.)
make: *** [Makefile:105: out/windows-amd64/crc.exe] Error 1
```

Not sure what the problem is

```
$ make vendor
go: github.com/crc-org/crc/v2/pkg/crc/machine imports
        github.com/crc-org/crc/pkg/crc/machine/config: github.com/crc-org/crc@v1.99.4: parsing go.mod:
        module declares its path as: github.com/code-ready/crc
                but was required as: github.com/crc-org/crc
make: *** [Makefile:79: vendor] Error 1
```

???

adrianriobo commented:
These are the results for this PR on windows https://crcqe-asia.s3.amazonaws.com/nightly/ocp/4.13.12/20231005/11-pro/qe-results/e2e-non-ux.xml `all green` except the known issue with the latest version for the pipelines operator

gbraad commented:
Want to incorporate 9P, but the current codebase for this lives in the PR https://github.com/containers/gvisor-tap-vsock/pull/280

cfergeau commented:
> Want to incorporate 9P, but the current codebase for this lives in the PR [containers/gvisor-tap-vsock#280](https://github.com/containers/gvisor-tap-vsock/pull/280)

The code from this gvisor-tap-vsock PR is short, and thus easy to add to the daemon when we need it. But there's also some guest code involved, which might be trickier https://github.com/mheon/libpod/commit/334e213662adcd71097df07e1090db53c03d15f7

gbraad commented:
@cfergeau it basically is the same/similar code as the 9p-ufs that I tested with earlier, so I am not so concerned. I am more curious if this will therefore be part of a library, or the changes for libpod remain seperate?

gbraad commented:
ref: #3870

cfergeau commented on pkg/crc/preflight/preflight_checks_windows.go:
```diff
@@ -166,19 +165,6 @@ func fixUserPartOfCrcUsersAndHypervAdminsGroup() error {
 	return errReboot
 }
 
-func checkIfHyperVVirtualSwitchExists() error {
```
Fwiw, it would be good to split the system mode networking removal on Windows in its own preparatory commit, this would make it easier to focus on/review the config/preflight changes.

gbraad replied:
Honestly, I do not even know if this functionality still works. I'll consider, but these checks are also performed when not even necessary (like the existence of the virtual switch) and therefore actually wrong.

cfergeau replied:
I've done this in https://github.com/cfergeau/crc/tree/libhvee, this removes even more code

cfergeau commented on pkg/drivers/libhvee/libhvee_windows.go:
```diff
@@ -0,0 +1,309 @@
+		return err
+	}
+
+	// Check that the user is an Administrator
```
Is this still needed without system networking?

gbraad replied:
Yes, as VMs can only be created as a Hyper-V Admin.

cfergeau commented on pkg/drivers/libhvee/libhvee_windows.go:
```diff
@@ -0,0 +1,309 @@
+				ms.DynamicMemoryEnabled = d.DynamicMemory
+				ms.VirtualQuantity = uint64(d.Memory)
+				ms.Reservation = 1024
+				ms.Limit = uint64(d.Memory)
```
I assume that the unit used for memory sizes are the same between crc and libhvee?

gbraad replied:
yes, they are.

cfergeau commented on pkg/drivers/libhvee/libhvee_windows.go:
```diff
@@ -0,0 +1,309 @@
+
+	log.Debugf("Machine: libhvee -> creating: done")
+
+	return nil
```
This was `return d.resizeDisk()` in the hyperv driver. This used to be needed to get the disk size correct when doing 
```
crc delete
crc config set disk-size 50
crc start
```
Are you sure it's no longer needed?

gbraad replied:
Will check. Might have been missed as resize scenario has mostly been untested.

Though, our testsuite passed here?

cfergeau replied:
I've tested this, and the `resizeDisk` call is required for initial disk resize to work. I've made the change.

cfergeau commented on pkg/drivers/libhvee/libhvee_windows.go:
```diff
@@ -0,0 +1,309 @@
+	}
+
+	log.Debugf("Machine: libhvee -> stop")
+	return vm.Stop()
```
Is this synchronous until the vm state is stopped? Or do we need to keep the `waitStopped` helper from the hyperv code?

cfergeau replied:
libhvee implementation is synchronous: https://github.com/containers/libhvee/blob/0698042e8c753a95c349a2c45f8a6b04a5aa2e3a/pkg/hypervctl/vm.go#L266-L276

cfergeau commented on pkg/drivers/libhvee/libhvee_windows.go:
```diff
@@ -0,0 +1,309 @@
+}
+
+func (d *Driver) GetIP() (string, error) {
+	return "", fmt.Errorf("")
```
Can this return a better error? This could be `return "", drivers.ErrNotSupported`
(`drivers` is 	`"github.com/crc-org/machine/libmachine/drivers"`)

cfergeau commented on pkg/drivers/libhvee/libhvee_windows.go:
```diff
@@ -0,0 +1,309 @@
+
+func (d *Driver) resizeDisk(newSize int64) error {
+	diskPath := d.getDiskPath()
+	out, err := cmdOut(fmt.Sprintf("@(Get-VHD -Path %s).Size", quote(diskPath)))
```
No API in libhvee for this and `Hyper-V\\Resize-VHD`?

gbraad replied:
No. Does not exist. Suggestion for improvement?

cfergeau replied:
Was about to suggest creating an issue in libhvee to track this, but it already exists https://github.com/containers/libhvee/issues/54

cfergeau replied:
Maybe you can add a link to the issue as a comment

gbraad replied:
Brent proposed something and is being reviewed.

cfergeau reviewed (commented):
I've rebased this and squashed the commits in https://github.com/cfergeau/crc/tree/libhvee
This looks like a fairly straightforward change, though I haven't tried running it on a Windows machine :)

gbraad commented:
This PR needs https://github.com/containers/libhvee/pull/66

cfergeau committed:
libhvee: Use libhvee APIs for disk resize

Disk handling API was added to libhvee since the first iteration of the
libhvee support, we can now make use of it.

Signed-off-by: Christophe Fergeau <cfergeau@redhat.com>

cfergeau commented:
Rebased again, and added support for disk resize in https://github.com/cfergeau/crc/tree/libhvee (untested)
Seeing that the branch is in github.com/crc-org, I'll just force push to it instead of keeping my own branch.

gbraad commented:
Rebasing... (experienced weird issues with branch but should be resolved now)

cfergeau committed:
windows: Remove system networking

Should return an error when trying to set the "system networking"
option. But I think this is already the case, or it's not possible to
choose this option in release builds?

Signed-off-by: Christophe Fergeau <cfergeau@redhat.com>

cfergeau commented:
This is still WIP, but I've tested that this works, including disk resize.
The 9p code needs more work which is described in the relevant commit log.

cfergeau commented:
Given https://github.com/crc-org/crc/issues/3870#issuecomment-2074870282 , I'd punt 9p support for now and work on merging the libhvee code with no changes to the file sharing code.

cfergeau commented:
> Given [#3870 (comment)](https://github.com/crc-org/crc/issues/3870#issuecomment-2074870282) , I'd punt 9p support for now and work on merging the libhvee code with no changes to the file sharing code.

Maybe we need to use https://github.com/aperezdc/9pfuse?tab=readme-ov-file or something similar.

gbraad commented:
> work on merging the libhvee code with no changes to the file sharing code.

Do you mean to target SMB or leave as in the current PR (no support)

cfergeau commented:
> > work on merging the libhvee code with no changes to the file sharing code.
> 
> Do you mean to target SMB or leave as in the current PR (no support)

Ideally, use the same SMB support as what we have now. I haven't checked if this PR removes some SMB support, but hopefully it's still there?

gbraad commented:
Originally the PR removed it, but part of it is there... just never tested. Note: we have no automated tests for this :-s.

anjannath commented:
Tested this PR on windows and the SMB based file sharing and it is working as expected, have to use the windows MSI installer, then the following config options needs to be set before `crc start`
```
$ crc config set host-network-access true
$ crc config set shared-dir-password <login_user_pass>
$ crc config set enable-shared-dir true
```

anjannath commented:
While testing i also noticed that error messages from the libhvee driver is a bit concise than the older hyperv driver, e.g

### from libhvee..
![Screenshot_win11_2024-05-21_14:26:15](https://github.com/crc-org/crc/assets/8885742/5bb41948-fb87-4d2b-89e0-a1ac8e7b0e0d)

---
### from hyperv driver
![Screenshot_win11_2024-05-21_14:44:07](https://github.com/crc-org/crc/assets/8885742/4746665a-2494-4c95-a006-e770dad73efc)

gbraad commented:
@anjannath this is a side effect of the WMI use. The powershell interface is more verbose, though also not always right. When WMI is used, you only get return codes back from the 'queries' and 'requests' that are performed. This would need to be a refinement in the libhvee library. Perhaps file an issue upstream for this?

gbraad commented:
/me tested and concludes: "works as expected"

gbraad commented:
Everlasting rebasing

cfergeau commented:
> Everlasting rebasing

Rebase done.

gbraad commented:
@lilyLuLiu @adrianriobo For this change, the proxy tests are failing. Do we have more information aroudn this?

adrianriobo commented:
@gbraad those are mainly failing due to timeout on the full suite, let me check downstream running only that one I will ack if it is green (and will increase the timeout for the suite here)

lilyLuLiu commented:
Downstream windows test passed for this PR.

adrianriobo commented:
~~:pray: hold till today mid day need to check something else~~

We are good to go, downstream proxy is green

adrianriobo reviewed (approved):
Full green on downstream windows env

cfergeau commented:
> Full green on downstream windows env

Didn't we wait too long though? We held this PR a while back as we did not want to land it at the same time as the switch to 4.16.
The next crc release could be 4.17-based, should we delay this again by one release?

praveenkumar commented:
> > Full green on downstream windows env
> 
> Didn't we wait too long though? We held this PR a while back as we did not want to land it at the same time as the switch to 4.16. The next crc release could be 4.17-based, should we delay this again by one release?

Since it is pass the pipeline, I don't think we should hold it any longer. I am merging this now.

cfergeau commented:
> Since it is pass the pipeline, I don't think we should hold it any longer. I am merging this now.

We held it last time because QE had concerns about their capacity to deal with 2 large changes at once, would have been nice to wait for their input before merging this..

adrianriobo commented:
No worries, I am fine with it, as @praveenkumar this already passed  (and actually has been passing for a while now). So I think we should be good.

gbraad commented:
Thank you all
